crates/ty_python_semantic/src/types/diagnostic.rs

```diff
@@ -1560,6 +1561,31 @@ declare_lint! {
     }
 }
 
+declare_lint! {
+    /// ## What it does
+    /// Detects `global` statements that don't match an explicit definition in the global scope.
+    ///
+    /// ## Why is this bad?
+    /// Function bodies with `global` statements can run in any order (or not at all), which makes
+    /// it hard to infer the types of globals without an explicit definition. Functions can also
+    /// read global variables without marking them `global`.
+    ///
+    /// ## Example
+    /// ```python
+    /// def f():
+    ///     global x  # unresolved global
+    ///     x = 42
+    ///
+    /// def g():
+    ///     print(x)  # unresolved reference
+    /// ```
+    pub(crate) static UNRESOLVED_GLOBAL = {
+        summary: "detects `global` statements with no definition in the global scope",
+        status: LintStatus::preview("1.0.0"),
+        default_level: Level::Error,
```

Comment:
since this is more about "this screws up the accuracy of our analysis elsewhere" rather than "this will raise an exception at runtime", I think `Warn` is a more appropriate default severity here:

```suggestion
        default_level: Level::Warn,
```